src/aws-cpp-sdk-core/source/utils/stream/ConcurrentStreamBuf.cpp

```diff
@@ -25,15 +27,106 @@ namespace Aws
                 setp(pbegin, pbegin + bufferLength);
             }
 
-            void ConcurrentStreamBuf::SetEof()
+            void ConcurrentStreamBuf::SetEofInput(Aws::IOStream* pStreamToClose)
             {
+                bool closeStream = false;
                 {
                     std::unique_lock<std::mutex> lock(m_lock);
-                    m_eof = true;
+                    m_eofInput = true;
+                }
+                FlushPutArea();
+                if (pStreamToClose)
+                {
+                    m_pStreamToClose = pStreamToClose;
+                    if (m_backbuf.empty())
+                    {
+                        closeStream = true;
+                    }
+                }
+                if (closeStream)
+                {
+                    CloseStream();
+                }
+                m_signal.notify_all();
+            }
+
+            void ConcurrentStreamBuf::CloseStream()
+            {
+                {
+                    std::unique_lock<std::mutex> lock(m_lock);
+                    m_eofOutput = true;
+                    if (m_pStreamToClose)
+                    {
+                        m_pStreamToClose->setstate(Aws::IOStream::eofbit);
+                        m_pStreamToClose = nullptr;
+                    }
                 }
                 m_signal.notify_all();
             }
 
+            bool ConcurrentStreamBuf::WaitForDrain(int64_t timeoutMs)
+            {
+                const auto waitStarted = std::chrono::steady_clock::now();
+
+                // timed FlushPutArea
```

Comment:
would maybe be useful to create a "timed action" utility template. you do three timed actions here and im sure we will need to do something like that in future again